### PR TITLE
Create output folder if it is missing

### PR DESCRIPTION
**Pull request by rvignesh89**
I've done a quick fix of what I wanted. It only affects those who enter an absolute path in the Output file parameter. Let me know what you think.

### CODE REVIEW

File: Code/Yahoo.Yui.Compressor.Build.MsBuild/CompressorTaskEngine.cs [hunk 1]
```diff
@@ -399,6 +400,17 @@ private bool SaveCompressedText(StringBuilder compressedText)
         {
             // Note: compressedText CAN be null or empty, so no check.
 
+            var folderNamePattern = @"(.*)\\(.*)\.(.*)$";
+            if (Regex.IsMatch(OutputFile, folderNamePattern))
```
Comment (PureKrome) [inline Code/Yahoo.Yui.Compressor.Build.MsBuild/CompressorTaskEngine.cs]: Can you please explain what this regex is trying to do. Please give some scenario examples. (eg. input & output).

**Comment by rvignesh89** [comment]
@PureKrome I used Regex previously to check if the path looks like

```
C:\somedir\nesterdir\output.js
```

But I realize that it doesn't really work for all cases. So I've made some changes to use Path.GetFullPath() to get the path of the output file. If it is a non existent directory it will be created. I've also added a test case which you can take a look.

File: Code/Yahoo.Yui.Compressor.Tests/JavaScriptCompressorTaskTests.cs [hunk 2]
```diff
@@ -182,6 +182,23 @@ public void An_Error_Is_Not_Logged_If_The_Output_File_Has_The_Same_Name_But_A_Di
...
+            compressor.SourceFiles = new ITaskItem[] { new TaskItem("Javascript Files\\SampleJavaScript1.js") };
+            compressor.OutputFile = "NewOutputDir\\Output.js";
+            if (Directory.Exists(Path.GetFullPath("NewOutputDir")))
+                Directory.Delete(Path.GetFullPath("NewOutputDir"), true);
+
+            // Act
+            compressor.Execute();
+
+            // Assert
+            Assert.That(BuildEngine.ContainsMessage(compressor.BuildEngine, ("Creating missing directory")));
```
Comment (freeranger) [inline Code/Yahoo.Yui.Compressor.Tests/JavaScriptCompressorTaskTests.cs]: The presence of the message is kind of irrelevant - shouldn't you be asserting that the directory was actually created?  That is after all the stated intent in the test name - that the directory is created if it does not exist, but you do not test this.
The fact that you wrote out a message saying the folder was created does not mean the folder was actually created.

File: Code/Yahoo.Yui.Compressor.Build.MsBuild/CompressorTaskEngine.cs [hunk 3]
```diff
@@ -395,10 +396,18 @@ private CompressionType GetCompressionTypeFor(FileSpec file)
...
         [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+        [SuppressMessage("ReSharper", "AssignNullToNotNullAttribute")]
         private bool SaveCompressedText(StringBuilder compressedText)
         {
             // Note: compressedText CAN be null or empty, so no check.
 
+            var outputDirectory = Path.GetDirectoryName(Path.GetFullPath(OutputFile));
+            if (!Directory.Exists(outputDirectory))
+            {
+                Log.LogMessage(string.Format("Creating missing directory {0}", outputDirectory));
```
Comment (freeranger) [inline Code/Yahoo.Yui.Compressor.Build.MsBuild/CompressorTaskEngine.cs]: I don't think you need the word "missing" in there.  It is the output directory so say "Creating output directory" instead

File: Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngine.cs [hunk 4]
```diff
@@ -8,29 +9,34 @@ public static class BuildEngine
...
+            return CheckForError("Errors", engine, error, false);
         }
 
-        private static bool CheckForError(IBuildEngine engine, string error, bool exists)
+        public static bool ContainsMessage(IBuildEngine engine, string message)
+        {
+            return CheckForError("Messages", engine, message, true);
+        }
+
+        private static bool CheckForError(string messageType, IBuildEngine engine, string searchString, bool exists)
```
Comment (freeranger) [inline Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngine.cs]: Since the method is now checking for Errors OR Messages, then "CheckForError" is no longer an appropriate name for the method - CheckForMessage probably makes more sense since a message could be an error or a regular message.
Also ideally the messageType would be an enum since there are only two valid values

File: Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngine.cs [hunk 5]
```diff
@@ -8,29 +9,34 @@ public static class BuildEngine
...
         {
             var buildEngine = engine as BuildEngineStub;
             if (buildEngine == null)
             {
                 Assert.Fail("Not a BuildEngineStub, cannot test with this");
             }
-            if (buildEngine.Errors != null && buildEngine.Errors.Count > 0)
+
+            var outputStrings = messageType == "Errors" ? buildEngine.Errors : buildEngine.Messages;
```
Comment (freeranger) [inline Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngine.cs]: rename outputStrings to messages

File: Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngine.cs [hunk 6]
```diff
@@ -8,29 +9,34 @@ public static class BuildEngine
...
+
+            if (outputStrings.Count > 0)
             {
-                foreach (var anError in buildEngine.Errors)
+                if (outputStrings.Any(anError => anError.StartsWith(searchString)))
```
Comment (freeranger) [inline Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngine.cs]: This is no longer an error so anError is a poor name for the variable.  With the previous refactor to use messages instead of outputStrings, change this line to:
`if (messages.Any(message => message.StartsWith(searchString)))`

File: Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngine.cs [hunk 7]
```diff
@@ -40,8 +46,8 @@ private static bool CheckForError(IBuildEngine engine, string error, bool exists
             }
 
             var sb = new StringBuilder();
-            sb.AppendLine(error + " not found.  Actual errors: ");
-            foreach (var anError in buildEngine.Errors)
+            sb.AppendLine(string.Format("{0} not found.  Actual {1}: ", searchString, messageType));
+            foreach (var anError in outputStrings)
```
Comment (freeranger) [inline Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngine.cs]: Again, use message instead of anError and messages instead of outputStrings

**Comment by rvignesh89** [comment]
@freeranger I'm going assert the test by checking if the directory exists. Realized this is way more easier than what I'm doing now.

File: Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngine.cs [hunk 8]
```diff
@@ -16,31 +18,29 @@ public static bool DoesNotContainError(IBuildEngine engine, string error)
             return CheckForError(engine, error, false);
         }
 
-        private static bool CheckForError(IBuildEngine engine, string error, bool exists)
+        private static bool CheckForError(IBuildEngine engine, string searchString, bool exists)
```
Comment (freeranger) [inline Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngine.cs]: My previous comment about this method still applies.

> Since the method is now checking for Errors OR Messages, then CheckForError is no longer an appropriate name for the method - CheckForMessage probably makes more sense since a message could be an error or a regular message.
Also ideally the messageType would be an enum since there are only two valid values.

But since you no longer test the output message then an alternative fix is to just revert you change completely - and leave it as a "CheckForError" method.
Please revert your change or make the changes I suggested

File: Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngineStub.cs [hunk 9]
```diff
@@ -27,6 +29,7 @@ public void LogWarningEvent(BuildWarningEventArgs e)
 
         public void LogMessageEvent(BuildMessageEventArgs e)
         {
+            Messages.Add(e.Message);
         }
```
Comment (freeranger) [inline Code/Yahoo.Yui.Compressor.Tests/TestHelpers/BuildEngineStub.cs]: Since you aren't checking messages any more, you can completely revert the changes to this file.
Try to keep your PR as tight as possible, doing the minimum necessary to adequately meet the requirement.

File: YUICompressor.sln.DotSettings [hunk 10]
```diff
@@ -222,7 +222,10 @@ II.2.12 &lt;HandlesEvent /&gt;&#xD;
 &lt;/Patterns&gt;&#xD;
 </s:String>
 	<s:String x:Key="/Default/Environment/Hierarchy/PsiConfiguration/LocationType/@EntryValue">TEMP_FOLDER</s:String>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpFileLayoutPatternsUpgrade/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EAddAccessorOwnerDeclarationBracesMigration/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateBlankLinesAroundFieldToBlankLinesAroundProperty/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateThisQualifierSettings/@EntryIndexedValue">True</s:Boolean>
```
Comment (freeranger) [inline YUICompressor.sln.DotSettings]: Please revert the changes to the DotSettings file.  We will maintain this

File: Code/Yahoo.Yui.Compressor.Tests/JavaScriptCompressorTaskTests.cs [hunk 11]
```diff
@@ -182,6 +182,23 @@ public void An_Error_Is_Not_Logged_If_The_Output_File_Has_The_Same_Name_But_A_Di
...
 
+        [Test]
+        public void If_OutputFile_Directory_Does_Not_Exist_It_Is_Created()
+        {
+            // Arange
+            var compressor = CreateCompressorTask();
+            compressor.SourceFiles = new ITaskItem[] { new TaskItem("Javascript Files\\SampleJavaScript1.js") };
+            compressor.OutputFile = "NewOutputDir\\Output.js";
+            if (Directory.Exists(Path.GetFullPath("NewOutputDir")))
+                Directory.Delete(Path.GetFullPath("NewOutputDir"), true);
```
Comment (freeranger) [inline Code/Yahoo.Yui.Compressor.Tests/JavaScriptCompressorTaskTests.cs]: Please check that the directory was actually deleted after you attempt the delete.  If it failed silently then your test will pass whether or not there exists any code to actually create the folder.
Also please use {} for all if statements.

**Comment by freeranger** [comment]
@rvignesh89 - a few tweaks and reverts for you.  Also can you please squash your commits as we would like to see a single commit in the history when we accept your PR and not the 1/2 dozen or so that you made in the process of making the changes for the PR.
This keeps the history of the project cleaner.

Thanks

File: Code/Yahoo.Yui.Compressor.Tests/JavaScriptCompressorTaskTests.cs (hunk 11 above)
Comment (rvignesh89) [inline Code/Yahoo.Yui.Compressor.Tests/JavaScriptCompressorTaskTests.cs]: Based on [Directory.Delete()](https://msdn.microsoft.com/en-us/library/fxeahc5f(v=vs.110).aspx)'s documentation I don't see a reason for it to fail silently. It is a fairly simple method call which throws an exception when it is unable to delete the directory which should fail the test case.

**Comment by rvignesh89** [comment]
@freeranger I noticed that Github has added a feature to [squash commits](https://github.com/blog/2141-squash-your-commits) when merging PR's. Do you think you can do that here? I've never personally used that though.

File: Code/Yahoo.Yui.Compressor.Tests/JavaScriptCompressorTaskTests.cs (hunk 11 above)
Comment (freeranger) [inline Code/Yahoo.Yui.Compressor.Tests/JavaScriptCompressorTaskTests.cs]: Ok, I was thinking of File.Delete which can silently not delete a file (e.g. it is in use with delete sharing on), perhaps directory deletion doesn't suffer from the same problem.
Forgot to add, can you delete the directory afterwards also?  It's always good to clean up after a test (suggest a try/catch to ensure the delete always happens) rather than having directories linger afterwards...

**Comment by freeranger** [comment]
@rvignesh89  Interesting - this is exactly a feature I've been looking for.  I want to keep my history on my fork of a project but have a single merge pulled into the master.
Ok leave your separate commits and I'll try this squash-commit to the master once you make the other changes...

**Comment by PureKrome** [comment]
Squashing commits during merge process is nice .. but .. it can leave your own repo-clones in a weird state or something. We (at our work) squash locally and push ... or force push if committing multiple times during the branch development, then squashing then force push.

YMMV :)

TL;DR; One commit per PR, plz. :+1:
